README.md
# react-bootstrap-button-loader
Bootstrap Button with Spinner feature

[![react-bootstrap-button-loader](https://github.com/yury-dymov/react-bootstrap-button-loader/raw/master/docs/gifs/demo.gif)](https://github.com/yury-dymov/react-bootstrap-button-loader)

[![npm version](https://img.shields.io/npm/v/react-bootstrap-button-loader.svg?style=flat)](https://www.npmjs.com/package/react-bootstrap-button-loader)
[![Downloads](http://img.shields.io/npm/dm/react-bootstrap-button-loader.svg?style=flat-square)](https://npmjs.org/package/react-bootstrap-button-loader)
[![Build Status](https://img.shields.io/travis/yury-dymov/react-bootstrap-button-loader/master.svg?style=flat)](https://travis-ci.org/yury-dymov/react-bootstrap-button-loader)
[![Coverage Status](https://coveralls.io/repos/github/yury-dymov/react-bootstrap-button-loader/badge.svg?branch=master)](https://coveralls.io/github/yury-dymov/react-bootstrap-button-loader?branch=master)

# Versioning
* v2 supports React Bootstrap v1.0.0+
* v1 supports React Bootstrap v0.x.x

# Demo
Demo and playground are available [here](https://yury-dymov.github.io/react-bootstrap-button-loader/)

# Usage Example
```
import Button from 'react-bootstrap-button-loader';

<Button loading={this.state.loading}>Press me!</Button>
```

# Configurable Props
*Note*: All props are optional.

### ~bsStyle: string, default: null~
Bootstrap style, supported values: `default`, `primary`, `success`, `info`, `warning`, `danger`, `link`.

*Deprecated in a favor of `variant`*

### disabled: boolean, default: false
Set this prop `true` to disable button.

*Note:* button in loading state is disabled and this behavior is not overridable even if `false` value is explicitly provided.

### loading: boolean, default: false
This prop controls Button loading state.

While loading, Button is disabled and icon provided via props is replaced with Spinner.

### icon: node, default: null
Buttons with icons are better!

Provided icon is shown if Button is not in a loading state. Otherwise Spinner is rendered.

### spinAlignment: string, default 'left'
Controls spinAlignment, supported values: 'left' and 'right'

### spinColor: string, default: '#fff'
Spinner color for loading state.

While white color used by default works fine for most cases, for different bootstrap themes and bsStyles it might be better to use other colors instead.

### variant: string, default: 'primary'
Bootstrap style, supported values: `primary`, ``secondary`, `success`, `info`, `warning`, `danger`, `link`, `light`, `dark`.


# License
MIT (c) Yury Dymov


CHANGELOG.md
### 17th June 2019 - Version 2.0.0
- Support for react-bootstrap 1.0.0 and deps update

### 17th June 2019 - Version 1.0.13
- Max version for react-bootstrap

### 15th June 2018 - Version 1.0.12
- Deps update

### 8th December 2017 - Version 1.0.11
- Fixed typo in default spinner style (https://github.com/yury-dymov/react-bootstrap-button-loader/pull/5)

### 17th October 2017 - Version 1.0.9
- Added prop `spinAlignment` to control spin alignment

### 19th May 2017 - Version 1.0.8
- Added `prop-types` package to support React 16+


test/ButtonLoader.spec.js
// We don't need to mutate props as ButtonLoader is stateless component

import React                      from 'react';
import chai, { expect }           from 'chai';
import chaiEnzyme                 from 'chai-enzyme';
import Enzyme, { shallow, mount, render } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import ButtonLoader, { Spinner }  from '../dist';

Enzyme.configure({ adapter: new EnzymeAdapter() });
chai.use(chaiEnzyme());

const icon = <span className='providedIcon'><svg xmlns="http://www.w3.org/2000/svg" width="20" height="14" viewBox="0 0 50 50"><path style={{fill: 'rgb(41, 128, 185)',textIndent:0,textAlign:'start',lineHeight: 'normal',textTransform: 'none',blockProgression: 'tb', InkscapeFontSpecification: 'Bitstream Vera Sans'}} d="M25,47.302l-0.64-0.533c-1.217-1.015-2.861-2.115-4.765-3.39C12.169,38.408,2,31.601,2,20C2,12.832,7.832,7,15,7 c3.896,0,7.542,1.734,10,4.699C27.458,8.734,31.104,7,35,7c7.168,0,13,5.832,13,13c0,11.601-10.169,18.408-17.595,23.379 c-1.904,1.274-3.548,2.375-4.765,3.39L25,47.302z" overflow="visible" enableBackground="accumulate" fontFamily="Bitstream Vera Sans" /></svg></span>;

describe('loading and disabled. If loading then button is disabled, disabled prop value might be any', () => {
  it('loading: true, disabled: undefined => disabled', () => {
    const component = mount(<ButtonLoader loading>Press me!</ButtonLoader>);

    expect(component.find('button')).to.be.disabled()
  });

  it('loading: true, disabled: true => disabled', () => {
    const component = mount(<ButtonLoader loading disabled>Press me!</ButtonLoader>);

    expect(component.find('button')).to.be.disabled();
  });

  it('loading: true, disabled: any => disabled', () => {
    const component = mount(<ButtonLoader loading disabled={false}>Press me!</ButtonLoader>);

    expect(component.find('button')).to.be.disabled();
  });
});

describe('disabled and loading. Button can be in loading state even if it was disabled', () => {
  it('disabled: true, loading: undefined => loading: false', () => {
    const component = render(<ButtonLoader disabled>Press me!</ButtonLoader>);

    expect(component.find('.loader')).to.have.length(0);
  });

  it('disabled: true, loading: false => loading: false', () => {
    const component = render(<ButtonLoader disabled loading={false}>Press me!</ButtonLoader>);

    expect(component.find('.loader')).to.have.length(0);
  });

  it('disabled: true, loading: true => loading: true', () => {
    const component = render(<ButtonLoader disabled loading>Press me!</ButtonLoader>);

    expect(component.find('.loader')).to.have.length(1);
  });
});

describe('Provided icon is rendered if not loading. Otherwise showing spinner', () => {
  it('icon: undefined, loading: false => icon: undefined', () => {
    const component = render(<ButtonLoader loading={false}>Press me!</ButtonLoader>);

    expect(component.find('.providedIcon')).to.have.length(0);
    expect(component.find('.loader')).to.have.length(0);
  });

  it('icon: undefined, loading: undefined => icon: undefined', () => {
    const component = render(<ButtonLoader>Press me!</ButtonLoader>);

    expect(component.find('.providedIcon')).to.have.length(0);
    expect(component.find('.loader')).to.have.length(0);
  });

  it('icon: <svg />, loading: false => icon: <svg />', () => {
    const component = render(<ButtonLoader icon={icon} loading={false}>Press me!</ButtonLoader>);

    expect(component.find('.providedIcon')).to.have.length(1);
    expect(component.find('.loader')).to.have.length(0);
  });

  it('icon: <svg />, loading: undefined => icon: <svg />', () => {
    const component = render(<ButtonLoader icon={icon}>Press me!</ButtonLoader>);

    expect(component.find('.providedIcon')).to.have.length(1);
    expect(component.find('.loader')).to.have.length(0);
  });

  it('icon: <svg />, loading: true => icon: spinner', () => {
    const component = render(<ButtonLoader icon={icon} loading={true}>Press me!</ButtonLoader>);

    expect(component.find('.providedIcon')).to.have.length(0);
    expect(component.find('.loader')).to.have.length(1);
  });

  it('icon: undefined, loading: true => icon: spinner', () => {
    const component = render(<ButtonLoader loading={true}>Press me!</ButtonLoader>);

    expect(component.find('.providedIcon')).to.have.length(0);
    expect(component.find('.loader')).to.have.length(1);
  });
});

describe('children prop is used', () => {
  it('passed as children prop', () => {
    const component = mount(<ButtonLoader children="Press me" />);

    expect(component.find('button')).to.have.html().match(/Press me/);
  });

  it('passed as xml', () => {
    const component = mount(<ButtonLoader>Press me</ButtonLoader>);

    expect(component.find('button')).to.have.html().match(/Press me/);
  });
});

describe('bsStyle, spinColor, style and className props are propagated', () => {

  it('bsStyle: success -> .btn-success', () => {
    const component = mount(<ButtonLoader bsStyle="success" children="Press me" />);

    expect(component.find('.btn-success')).to.have.length(1);
  });

  it('expect Spinner to have color #333', () => {
    const component = mount(<ButtonLoader  children="Press me" loading spinColor="#333" />);

    expect(component.find(Spinner).props().spinColor).to.equal('#333');
  });

  it('className: "temp"', () => {
    const component = render(<ButtonLoader className="temp" />);

    expect(component).to.have.attr('class').match(/temp/);
  });

  it('style: {{ zIndex: "-9999" }}', () => {
    const component = render(<ButtonLoader style={{ zIndex: '-9999' }} />);

    expect(component).to.have.attr('style').match(/\-9999/);
  });
});


demo/src/index.js
import React, { Component } from 'react';
import { render } from 'react-dom';
import Button from 'react-bootstrap-button-loader';

class App extends Component {
  static svgIcon() {
    return (
      <span>
        <svg xmlns="http://www.w3.org/2000/svg" width="20" height="14" viewBox="0 0 50 50">
          <path
            style={{
              fill: 'rgb(41, 128, 185)',
              textIndent:0,
              textAlign:'start',
              lineHeight: 'normal',
              textTransform: 'none',
              blockProgression: 'tb',
              InkscapeFontSpecification: 'Bitstream Vera Sans',
            }}
            d="M25,47.302l-0.64-0.533c-1.217-1.015-2.861-2.115-4.765-3.39C12.169,38.408,2,31.601,2,20C2,12.832,7.832,7,15,7 c3.896,0,7.542,1.734,10,4.699C27.458,8.734,31.104,7,35,7c7.168,0,13,5.832,13,13c0,11.601-10.169,18.408-17.595,23.379 c-1.904,1.274-3.548,2.375-4.765,3.39L25,47.302z"
            overflow="visible"
            enableBackground="accumulate"
            fontFamily="Bitstream Vera Sans"
          />
        </svg>
      </span>
    );
  }

  constructor() {
    super();

    this.handleClick = this.handleClick.bind(this);

    this.state = {
      disabled: false,
      forceLoading: false,
      loading: 0,
      showIcon: true,
      spinColor: '#fff',
      style: 'default',
    };
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  handleClick() {
    this.setState({ loading: 5 });
    this.interval = setInterval(() => {
      const nextLoading = this.state.loading - 1;

      if (nextLoading <= 0) {
        clearInterval(this.interval);
      }

      this.setState({ loading: Math.max(nextLoading, 0) });

    }, 1000);
  }

  render() {
    const bsButtons = ['default', 'primary', 'success', 'info', 'warning', 'danger', 'link'].map((bsStyle) => (
      <Button
        bsStyle={bsStyle}
        disable={this.state.bsStyle === bsStyle}
        onClick={() => { this.setState({ bsStyle }); }}
        style={{ marginRight: '10px' }}
      >
        {bsStyle}
      </Button>
    ));

    const spinColorButtons = ['#fff', '#444'].map((color) => (
      <Button
        disabled={this.state.spinColor === color}
        onClick={() => this.setState({ spinColor: color })}
        style={{ marginRight: '10px' }}
      >
        {color}
      </Button>
    ));

    return (
      <div>
        <div>
          <Button
            bsStyle={this.state.bsStyle}
            disabled={this.state.disabled}
            icon={this.state.showIcon ? this.constructor.svgIcon() : null}
            loading={this.state.loading || this.state.forceLoading}
            onClick={this.handleClick}
            spinColor={this.state.spinColor}
          >
            Press Me!
          </Button>
          {!!this.state.loading && <p>I am loading for {this.state.loading} more seconds</p>}
        </div>
        <div style={{ marginTop: '40px' }}>
          <hr />
          <h1>Configurable props</h1>
          <h3>bsStype</h3>
          {bsButtons}
          <h3>disabled</h3>
          <Button onClick={() => this.setState({ disabled: !this.state.disabled })}>{this.state.disabled ? 'enable' : 'disabled'}</Button>
          <h3>loading</h3>
          <Button disabled={this.state.loading} onClick={() => this.setState({ forceLoading: !this.state.forceLoading })}>{this.state.forceLoading ? 'stop' : 'start'}</Button>
          <h3>icon</h3>
          <Button onClick={() => this.setState({ showIcon: !this.state.showIcon })}>{this.state.showIcon ? 'hide' : 'show'}</Button>
          <h3>spinColor</h3>
          {spinColorButtons}
        </div>
      </div>
    );
  }
}

render(<App />, document.getElementById('app'));
